helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -144,19 +314,91 @@ public ZNRecord update(ZNRecord current) {
       // If no one owns the lock, allow the update
       // If the user is the current lock owner, allow the update
       LockInfo curLockInfo = new LockInfo(current);
-      if (!(System.currentTimeMillis() < curLockInfo.getTimeout()) || isCurrentOwner()) {
+      if (System.currentTimeMillis() > curLockInfo.getTimeout() || isCurrentOwner(curLockInfo)) {
         return _record;
       }
-      // For users who are not the lock owner and try to do an update on a lock that is held by
-      // someone else, exception thrown is to be caught by data accessor, and return false for
-      // the update
+
+      // higher priority lock request will preempt current lock owner that is with lower priority
+      if (!isCurrentOwner(curLockInfo) && _priority > curLockInfo.getPriority() && curLockInfo
+          .getRequestorId().equals("NONE")) {
+        // update lock Znode with requestor information
+        ZNRecord newRecord = composeNewRequestorRecord(curLockInfo, _record);
+        _isPending = true;
+        _pendingTimeout = _waitingTimeout > curLockInfo.getCleanupTimeout() ? curLockInfo.getCleanupTimeout()
+                : _waitingTimeout;
+        return newRecord;
+      }
+
+      // If the requestor field is not empty, and the coming lock request has a even higher
+      // priority. The new request will replace current requestor field of the lock
+      if (!isCurrentOwner(curLockInfo) && _priority > curLockInfo.getPriority() && !curLockInfo
+          .getRequestorId().equals("NONE") && _priority > curLockInfo.getRequestorPriority()) {
```

Comment:
This could be a sub branch of previous one since the previous two checks are same. Can we combine them?